File: accounting-system-web-project/nexa-erp-frontend/src/app/features/audit/services/audit-log.service.ts
import { HttpClient, HttpParams } from '@angular/common/http';
import { Injectable } from '@angular/core';
import { Observable } from 'rxjs';
import { APP_CONFIG } from '../../../core/config/app.config';
import { ApiResponse } from '../../../core/models/api-response.model';
import { PageResponse } from '../../../core/models/page.model';
import {
  AuditLog,
  AuditTimelineEntityName,
  AuditTimelineItem,
} from '../models/audit-log.model';

@Injectable({
  providedIn: 'root',
})
export class AuditLogService {
  private readonly baseUrl = `${APP_CONFIG.apiUrl}/audit-logs`;

  constructor(private http: HttpClient) {}

  // All logs for an entity type, e.g. all INVOICE logs
  getEntityLogs(
    entityName: string,
    page = 0,
    size = 20,
  ): Observable<ApiResponse<PageResponse<AuditLog>>> {
    const params = new HttpParams().set('page', page).set('size', size);
    return this.http.get<ApiResponse<PageResponse<AuditLog>>>(
      `${this.baseUrl}/entity/${entityName}`,
      { params },
    );
  }

  // Full history of one specific record, e.g. INVOICE #42
  getEntityHistory(
    entityName: string,
    entityId: number,
    page = 0,
    size = 20,
  ): Observable<ApiResponse<PageResponse<AuditLog>>> {
    const params = new HttpParams().set('page', page).set('size', size);
    return this.http.get<ApiResponse<PageResponse<AuditLog>>>(
      `${this.baseUrl}/entity/${entityName}/${entityId}`,
      { params },
    );
  }

  getEntityTimeline(
    entityName: AuditTimelineEntityName,
    entityId: number,
    page = 0,
    size = 20,
  ): Observable<ApiResponse<PageResponse<AuditTimelineItem>>> {
    const params = new HttpParams().set('page', page).set('size', Math.min(size, 20));
    return this.http.get<ApiResponse<PageResponse<AuditTimelineItem>>>(
      `${this.baseUrl}/entity/${entityName}/${entityId}/timeline`,
      { params },
    );
  }

  // Everything a specific user has done
  getUserActivity(
    userId: number,
    page = 0,
    size = 20,
  ): Observable<ApiResponse<PageResponse<AuditLog>>> {
    const params = new HttpParams().set('page', page).set('size', size);
    return this.http.get<ApiResponse<PageResponse<AuditLog>>>(`${this.baseUrl}/user/${userId}`, {
      params,
    });
  }
}

File: backend-for-all/backend/src/main/java/com/nexaerp/audit/AuditLogService.java
package com.nexaerp.audit;

import com.nexaerp.audit.dto.AuditTimelineItemDto;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;

public interface AuditLogService {
    // Save a log entry
    void log(AuditAction action, String entityName, Long entityId,
             String oldValue, String newValue);

    // Get history of a specific record
    Page<AuditLog> getEntityHistory(String entityName, Long entityId, Pageable pageable);

    // Get activity of a specific user
    Page<AuditLog> getUserActivity(Long userId, Pageable pageable);

    // Get all logs for an entity type
    Page<AuditLog> getEntityLogs(String entityName, Pageable pageable);

    // Get safe timeline data for a supported transaction record
    Page<AuditTimelineItemDto> getEntityTimeline(
            String entityName,
            Long entityId,
            Pageable pageable
    );
}
